### ISSUE TITLE
how to deal with empty proposals for mask branch when training?

### ISSUE THREAD

**Issue by zimenglan-sysu-512** [facebookresearch/maskrcnn-benchmark#497]
i find that if the proposals is empty when train the mask branch, the training procedure will hang on. how to deal with this cases?

**Comment by fmassa** [comment]
This is due to a different codepath being executed in the backwards, and some hook not being triggered properly in the `distributed` package.

I think the right solution here would be to make `conv` , `batch_norm` / etc support empty batch sizes natively.

Another solution I believe would work is to use the legacy distributed backend.

**Comment by Sreehari-iitm** [comment]
Experiencing the same issue.
![capture](https://user-images.githubusercontent.com/42383260/53679785-67593900-3cf7-11e9-958f-409e8617e409.PNG)

Training stops when mask proposals are empty. Any suggestions, to deal with this case?

**Comment by zimenglan-sysu-512** [comment]
hi @fmassa 
u are right,it needs to make `conv`, `bn`, `gn` / etc support empty batchsizes, so i follow `Conv2d` to encapsulate `GN` like this:
```
class GroupNorm(torch.nn.GroupNorm):
    def forward(self, x):
        if x.numel() > 0:
            return super(GroupNorm, self).forward(x)
        # get output shape
        output_shape = x.shape
        return _NewEmptyTensorOp.apply(x, output_shape)
```
but it does not work.

hi @Sreehari-iitm 
a simple but not nice method is that we make some fake proposals to forward the mask branch and then make zero for the gradients when bp.

**Comment by Sreehari-iitm** [comment]
hi @zimenglan-sysu-512 
I did this expecting the network to improvise after few iterations that it no more gives empty mask proposals, but the network was still giving empty proposals in some iterations. I think there should be something to penalize empty proposals so that it improves over time. But still didn't figure out a way to do that

**Comment by zimenglan-sysu-512** [comment]
hi @Sreehari-iitm 
maybe try [this in mask scoring rcnn ](https://github.com/zjhuang22/maskscoring_rcnn/blob/master/maskrcnn_benchmark/modeling/roi_heads/maskiou_head/maskiou_head.py#L38)to deal with empty proposals for mask branch.

**Comment by Sreehari-iitm** [comment]
@zimenglan-sysu-512 @fmassa 
I tried training the model with mask head turned off only to predict bounding boxes, and it gave good results on my custom dataset. However when i turned on the mask head it is continuously giving empty proposals for most images.

```
def polypoints(mask):
    mask_list = []
    cnts = cv2.findContours(mask.copy().astype(np.uint8), cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
    cnts = imutils.grab_contours(cnts)
    cnts = sorted(cnts, key = cv2.contourArea, reverse = True)
    cnt = cnts[0]
    peri = cv2.arcLength(cnt, True)
    approx = cv2.approxPolyDP(cnt, 0.005 * peri, True)
    approx = approx.reshape(-1).tolist()
    mask_list.append(approx)
    return mask_list

```
This is how I'm converting the binary mask image to polygon format , which is used in coco.Any idea why this is happening?
Thank you
